### PR TITLE
Window support

### PR DESCRIPTION
**Pull request by hbxie**

### CODE REVIEW

File: app/templates/components/window-input.hbs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  <div class="emit-type-options">
+    {{#radio-button radioId="time-based" value=EMIT_TYPES.TIME groupValue=emitType changed=(action "changeEmitType" EMIT_TYPES.TIME) disabled=disabled}}
+      {{EMIT_TYPES.TIME}}
+    {{/radio-button}}
+    {{#radio-button radioId="record-based" value=EMIT_TYPES.RECORD groupValue=emitType changed=(action "changeEmitType" EMIT_TYPES.RECORD) disabled=recordBasedWindowDisabled}}
+      {{EMIT_TYPES.RECORD}}
+    {{/radio-button}}
+  </div>
+
+  <div class="subsection-header">Fill Your Emit frequency</div>
```
Comment (akshaisarma) [inline app/templates/components/window-input.hbs]: Emit Frequency

File: config/env-settings.json [hunk 2]
```diff
@@ -28,6 +28,7 @@
       "distributionQuantileEnd": 1,
       "distributionQuantileIncrement": 0.1,
       "queryTimeoutSecs": 3,
+      "timeWindowEmitFrequncyMinSecs": 1,
```
Comment (akshaisarma) [inline config/env-settings.json]: Can we call this windowEmitFrequencyMinSecs? Note typo in Frequency

File: tests/helpers/mocked-api.js [hunk 3]
```diff
@@ -49,7 +50,7 @@ export default EmberObject.extend({
     let dataArray = this.get('dataArray');
     if (onStompMessage && !isEmpty(dataArray)) {
       let length = dataArray.length;
-      dataArray.forEach((data, i) => {
+      dataArray.forEach((data, i) => next(() => {
```
Comment (akshaisarma) [inline tests/helpers/mocked-api.js]: New line and new level of indentation?

File: tests/integration/components/window-input-test.js [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+    await render(hbs `{{window-input addWindow=mockAddWindow query=mockQuery}}`);
+    assert.equal(this.element.querySelectorAll('.add-button').length, 1);
+
+    await click('.add-button');
+    assert.equal(this.element.querySelectorAll('.add-button').length, 0);
+    assert.equal(this.element.querySelectorAll('.remove-button').length, 1);
+  });
+
+  test('it closes window', async function(assert) {
```
Comment (akshaisarma) [inline tests/integration/components/window-input-test.js]: it removes window?

File: tests/unit/validators/window-emit-frequency-test.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    let mockModel = EmberObject.create({
+      emitType: EMIT_TYPES.get('TIME'),
+      query: query
+    });
+    let expected = 'The window emit frequency should not be longer than the query duration (20 seconds)';
+    assert.equal(validator.validate(21, null, mockModel), expected);
+    assert.ok(validator.validate(20, null, mockModel));
+  });
+
+  test('it does not let you subceed the minimum of emit frequency when it is time based window', function(assert) {
```
Comment (akshaisarma) [inline tests/unit/validators/window-emit-frequency-test.js]: ...exceed the minimum emit frequency...?

File: tests/unit/validators/window-emit-frequency-test.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+          timeWindowEmitFrequncyMinSecs: 10
+        }
+      }
+    });
+    let expected = 'The maintainer has configured Bullet to support a minimum of 10 for emit frequency';
+    assert.equal(validator.validate(9, null, mockModel), expected);
+    assert.ok(validator.validate(10, null, mockModel));
+  });
+
+  test('it is always validated when it is record based window', function(assert) {
```
Comment (akshaisarma) [inline tests/unit/validators/window-emit-frequency-test.js]: ...it successfully validates when...?

File: app/styles/components/window-input.scss [hunk 7]
```diff
@@ -0,0 +1,74 @@
+.window-input {
+  $border-rectangle-color: #DCDEE2;
```
[the remark this answers is not in the document]
Comment (hbxie) [reply app/styles/components/window-input.scss]: extracted some common section to query-input.scss and then output-data-input.scss and window-input.scss extend from there.

File: app/components/window-input.js [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+  // they are independent of the original property. This means they are only used on initial component render.
+  emitType: computed('query.window.emitType', function() {
+    return this.get('query.window.emitType');
+  }),
+
+  includeType: computed('query.window.includeType', function() {
+    return this.get('query.window.includeType');
+  }),
+
+  aggregationTypeChanged: observer('query.aggregation.type', function() {
```
[the remark this answers is not in the document]
Comment (hbxie) [reply app/components/window-input.js]: Moved this logic to isWindowless computed property.

File: app/models/result.js [hunk 9]
```diff
@@ -4,30 +4,39 @@
...
   pivotOptions: DS.attr('string'),
+  querySnapshot: DS.attr(),
+
+  // Removed it after result page has been changed.
+  metadata: computed('segments.[]', function() {
+    let segments = this.get('segments').toArray();
+    if (isEmpty(segments)) {
+      return null;
+    }
+    return segments.objectAt(segments.length - 1).get('metadata');
```
[the remark this answers is not in the document]
Comment (akshaisarma) [reply app/models/result.js]: Really? It's just a computed function that does exactly what you do here:
https://github.com/emberjs/ember.js/blob/v3.1.1/packages/ember-runtime/lib/mixins/array.js#L287

[the remark this answers is not in the document]
Comment (hbxie) [reply app/models/result.js]: It turns out the warning is caused by something else. Changed it to use lastObject.

File: app/services/query-manager.js [hunk 10]
```diff
@@ -150,6 +161,51 @@ export default Service.extend({
...
+      query.set('window', null);
+      return query.save().then(() => {
+        return window.destroyRecord();
+      });
+    });
+  },
+
+  addWindow(query) {
+    return query.get('window').then(window => {
+      if (!isEmpty(window)) {
```
Comment (akshaisarma) [inline app/services/query-manager.js]: Do we need this check? Can this be called with an existing window? It feels like the caller should be the one doing the checking if necessary

File: app/services/query-manager.js [hunk 11]
```diff
@@ -150,6 +161,51 @@ export default Service.extend({
...
+        type: window.get('include.type')
+      });
+      this.setIfNotEmpty(include, 'type', includeType);
+      window.set('emit', emit);
+      window.set('include', include);
+      return window.save();
+    });
+  },
+
+  deleteWindow(query) {
```
Comment (akshaisarma) [inline app/services/query-manager.js]: Move to the delete function section below

File: app/services/stomp-websocket.js [hunk 12]
```diff
@@ -27,28 +29,30 @@ export default Service.extend({
...
+  makeStompMessageHandler(stompClient, context) {
     return payload => {
       let message = JSON.parse(payload.body);
       if (!isEqual(message.type, ACK_TYPE)) {
         if (isEqual(message.type, COMPLETE_TYPE)) {
-          stompClient.disconnect();
+          this.get('querier').cancel();
         }
-        successHandler(JSON.parse(message.content), context);
+        context.get('queryManager').addSegment(context.get('result'), JSON.parse(message.content));
```
Comment (akshaisarma) [inline app/services/stomp-websocket.js]: We should not expect context to have a queryManager. We should inject it here ourselves. It is also weird that this knows there will be a result variable in context.

Should we pass in a messageHandler like we do a successHandler and an errorHandler?

File: app/styles/components/window-input.scss [hunk 13]
```diff
@@ -0,0 +1,56 @@
+.window-input {
+  .add-button {
+    @extend %add-button;
+    margin-bottom: 10px;
+    i {
+      padding-right: 5px;
+    }
+    &.add-group, &.add-metric {
```
Comment (akshaisarma) [inline app/styles/components/window-input.scss]: This doesn't exist

File: app/styles/components/window-input.scss [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+  }
+  .validated-input {
+    .labeled-input {
+      padding-right: 0;
+      width: 150px;
+      margin-left: -5px;
+      margin-right: 0;
+    }
+
+    .error-tooltip-link {
```
Comment (akshaisarma) [inline app/styles/components/window-input.scss]: This is not applicable here.

File: app/styles/components/window-input.scss [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+    @extend %ember-radio-button;
+    display: inline-block;
+    margin-right: 30px;
+  }
+
+  .subsection-header {
+    @extend %subsection-header;
+  }
+
+.row {
```
Comment (akshaisarma) [inline app/styles/components/window-input.scss]: This can also be abstracted no?

File: app/validators/valid-window.js [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+ *  Licensed under the terms of the Apache License, Version 2.0.
+ *  See the LICENSE file associated with the project for terms.
+ */
+import { isEqual } from '@ember/utils';
+import BaseValidator from 'ember-cp-validations/validators/base';
+import { AGGREGATIONS } from 'bullet-ui/models/aggregation';
+import { EMIT_TYPES, INCLUDE_TYPES } from 'bullet-ui/models/window';
+
+const ValidWindow = BaseValidator.extend({
+  validate(value, options, model) {
```
Comment (akshaisarma) [inline app/validators/valid-window.js]: s/RAW/Raw?

File: tests/helpers/mocked-query.js [hunk 17]
```diff
@@ -60,15 +60,39 @@ export default EmberObject.extend({
...
     });
-    A(['filter', 'projections', 'aggregation', 'results']).forEach(this.topLevelPropertyAsPromise, this);
+    A(['filter', 'projections', 'aggregation', 'results', 'window']).forEach(this.topLevelPropertyAsPromise, this);
+  },
+
+  isWindowless: computed('_window', function() {
+    return isEmpty(this.get('_window'));
+  }),
+
+  setWindow(emitType, emitEvery, includeType) {
```
Comment (akshaisarma) [inline tests/helpers/mocked-query.js]: addWindow. Can you move both these methods to below the other add* methods?

File: app/models/query.js [hunk 18]
```diff
@@ -115,6 +128,16 @@ export default DS.Model.extend(Validations, {
...
   }),
 
+  windowSummary: computed('isWindowless', 'window.{emit.type,emit.every,include.type}', function() {
+    if (this.get('isWindowless')) {
+      return 'None';
+    }
+    let emitType = this.get('window.emit.type');
+    let emitEvery = this.get('window.emit.every');
+    let includeType = this.get('window.include.type');
+    return `Every ${emitEvery} ${isEqual(emitType, EMIT_TYPES.get('TIME')) ? 'seconds' : 'records'} ${isEqual(includeType, INCLUDE_TYPES.get('ALL')) ? ',Cumulative' : ''}`;
```
[the remark this answers is not in the document]
Comment (hbxie) [reply app/models/query.js]: Done.

File: app/routes/query.js [hunk 19]
```diff
@@ -11,37 +11,31 @@ export default Route.extend({
...
     fireQuery() {
-      this.store.findRecord('query', this.paramsFor('query').query_id).then(query => {
-        let request = this.get('querier').send(query, this.resultHandler, this.errorHandler, this);
-        // The low level XMLHTTPRequest
-        this.set('pendingRequest', request);
+      this.get('queryManager').addResult(this.paramsFor('query').query_id).then(result => {
+        this.store.findRecord('query', this.paramsFor('query').query_id).then(query => {
+          this.set('result', result);
+          let handlers = {
+            successHandler: this.resultHandler,
```
[the remark this answers is not in the document]
Comment (hbxie) [reply app/routes/query.js]: Done.

File: app/services/stomp-websocket.js [hunk 20]
```diff
@@ -27,46 +29,47 @@ export default Service.extend({
     return this.get('settings.queryStompResponseChannel');
   }),
 
-  makeStompMessageHandler(stompClient, successHandler, context) {
+  makeStompMessageHandler(stompClient, handlers, context) {
     return payload => {
       let message = JSON.parse(payload.body);
```
[the remark this answers is not in the document]
Comment (hbxie) [reply app/services/stomp-websocket.js]: Done.

File: app/templates/components/window-input.hbs [hunk 21]
```diff
@@ -0,0 +1,45 @@
+{{!--
+  Copyright 2018, Yahoo Inc.
+  Licensed under the terms of the Apache License, Version 2.0.
+  See the LICENSE file associated with the project for terms.
+ --}}
+{{#if isWindowless}}
+  <div class="subsection-header">No Window</div>
+  <button {{action "addWindow" }} disabled= {{disabled}} class="btn btn-primary btn-xs add-button">
```
[the remark this answers is not in the document]
Comment (hbxie) [reply app/templates/components/window-input.hbs]: Done.

File: app/templates/components/window-input.hbs [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+    <i class="glyphicon glyphicon-plus"></i>
+    Window
+  </button>
+{{else}}
+  <div class="btn-group pull-right">
+    <button {{action "removeWindow" }} disabled= {{disabled}} class="btn btn-xs btn-danger remove-button">
```
[the remark this answers is not in the document]
Comment (hbxie) [reply app/templates/components/window-input.hbs]: Done.

**Pull request merged**